### PR TITLE
Remove constness from `ImplSource::Param`

### PR DESCRIPTION
**Pull request by fee1-dead**

### CODE REVIEW

File: compiler/rustc_ast_lowering/src/lib.rs [hunk 1]
```diff
@@ -2495,14 +2495,27 @@ impl<'hir> GenericArgsCtor<'hir> {
 
         let id = lcx.next_node_id();
         let hir_id = lcx.next_id();
+
+        let constness = match constness {
```
Comment (cjgillot) [inline compiler/rustc_ast_lowering/src/lib.rs:2499]: We already know that `constness` is `Yes` from a few lines above.

File: tests/ui/rfcs/rfc-2632-const-trait-impl/tilde_const_on_impl_bound.stderr [hunk 2]
```diff
@@ -1,11 +1,12 @@
...
+error[E0308]: mismatched types
+  --> $DIR/tilde_const_on_impl_bound.rs:14:9
    |
 LL |         self.0.foo()
-   |                ^^^^^
+   |         ^^^^^^^^^^^^ expected `host`, found `true`
    |
-   = note: calls in constant functions are limited to constant functions, tuple structs and tuple variants
+   = note: expected constant `host`
+              found constant `true`
```
Comment (cjgillot) [inline tests/ui/rfcs/rfc-2632-const-trait-impl/tilde_const_on_impl_bound.stderr:8]: The internal name `host` leaks.

Comment (fee1-dead) [reply tests/ui/rfcs/rfc-2632-const-trait-impl/tilde_const_on_impl_bound.stderr:8]: This is a known issue, there's a FIXME somewhere for improving this but I haven't had time to work on it yet

File: compiler/rustc_ast_lowering/src/lib.rs (hunk 1 above)
Comment (fee1-dead) [reply compiler/rustc_ast_lowering/src/lib.rs]: Yep, fixed.

**Review by davidtwco — APPROVED** [review]

**Pull request merged**